File: src/Hangman/Hangman.java
package Hangman;

public class Hangman {
    private String word;
    private int max_guesses;
    // takes less space than a list of `char`s
    private String correct_guesses = "";
    private String wrong_guesses = "";

    // create a new Hangman instance
    public Hangman(String word, int max_guesses) {
        this.word = word;
        this.max_guesses = max_guesses;
    }

    // converts the string to lowercase, lets users not care about capitalization
    public static Hangman with_ignore_case(String word, int max_guesses) {
        return new Hangman(word.toLowerCase(), max_guesses);
    }

    // guesses a letter, returning false if it is not in the word
    public boolean guess_letter(String letter) {
        // validate
        if (letter.length() != 1 || this.wrong_guesses.length() >= this.max_guesses) {
            return false;
        }

        // if we've already guessed that it is in the word, then we don't need to care
        if (this.correct_guesses.contains(letter)) {
            return true;
        }

        if (this.wrong_guesses.contains(letter)) {
            return false; // it'll show that it is wrong, but we have already guessed this
        }

        final boolean contains = this.word.contains(letter);

        if (contains) {
            correct_guesses += letter;
        } else {
            wrong_guesses += letter;
        }

        return contains;
    }

    // guess a letter without caring about casing. this will only work if we use Hangman.with_ignore_case
    public boolean guess_letter_ignore_case(String letter) {
        return this.guess_letter(letter.toLowerCase());
    }

    // get the currently revealed string, replacing unknown chars with a red '*'
    public String revealed() {
        String out = "";

        // walk tge strubg
        for (int i = 0; i < this.word.length(); i++) {
            // get current character
            final String c = this.word.substring(i, i + 1);
            // if we have it, make it a green version of the letter, terminal magic
            if (correct_guesses.contains(c)) {
                out += "\u001b[32m" + c + "\u001b[0m";
            } else {
                out += "\u001b[31m*\u001b[0m";
            }
        }

        return out;
    }

    // list all of our guesses
    public String guesses() {
        String out = "";

        // list of correct letters, in green
        for (int i = 0; i < this.correct_guesses.length(); i++) {
            final String c = this.correct_guesses.substring(i, i + 1);
            out += "\u001b[32m" + c + "\u001b[0m ";
        }

        // wrong letters, in red
        for (int i = 0; i < this.wrong_guesses.length(); i++) {
            final String c = this.wrong_guesses.substring(i, i + 1);
            out += "\u001b[31m" + c + "\u001b[0m ";
        }

        // trim extra spaces at the end, since we always add ' ' at the end of each letter
        while (out.endsWith(" ")) {
            out = out.substring(0, out.length() - 1);
        }

        return out;
    }

    public boolean can_guess() {
        // only count wrong_guesses, correct guesses do not count toward the guess count
        return this.wrong_guesses.length() < this.max_guesses;
    }

    public int guesses_left() {
        return this.max_guesses - this.wrong_guesses.length();
    }

    // checking if the player has won
    public boolean has_revealed_word() {
        for (int i = 0; i < this.word.length(); i++) {
            final String c = this.word.substring(i, i + 1);
            if (!correct_guesses.contains(c)) {
                return false;
            }

            // no else here, wait until the entire string is done to ensure that we have it all
        }

        return true;
    }

    public String get_word() {
        return this.word;
    }

    // not the same as correct_guesses.length(), this counts duplicate letters ('cc' would be counted as 2, not 1)
    public int get_found_count() {
        int count = 0;
        for (int i = 0; i < this.word.length(); i++) {
            final String c = this.word.substring(i, i + 1);
            if (correct_guesses.contains(c)) {
                count++;
            }
        }

        return count;
    }

}

File: rust/src/hangman.rs
pub struct Hangman {
    word: String,
    max_guesses: u32,
    correct_guesses: String,
    wrong_guesses: String,
}

impl Hangman {
    pub fn new(word: String, max_guesses: u32) -> Self {
        Self {
            correct_guesses: String::new(),
            max_guesses,
            word,
            wrong_guesses: String::new(),
        }
    }

    pub fn with_ignore(word: String, max_guesses: u32) -> Self {
        Self::new(word.to_lowercase(), max_guesses)
    }

    pub fn guess_letter(&mut self, letter: String) -> bool {
        if letter.len() != 1 || self.wrong_guesses.len() >= (self.max_guesses as usize) {
            return false;
        }

        if self.correct_guesses.contains(&letter) {
            return true;
        }

        if self.wrong_guesses.contains(&letter) {
            return false;
        }

        let contains = self.word.contains(&letter);

        if contains {
            self.correct_guesses += &letter;
        } else {
            self.wrong_guesses += &letter;
        }

        contains
    }

    pub fn guess_letter_ignore_case(&mut self, letter: String) -> bool {
        self.guess_letter(letter.to_lowercase())
    }

    pub fn revealed(&self) -> String {
        let mut out = String::new();

        for i in 0..self.word.len() {
            let c = &self.word[i..i + 1];

            if self.correct_guesses.contains(c) {
                out += "\u{001b}[32m";
                out += c;
                out += "\u{001b}[0m";
            } else {
                out += "\u{001b}[31m*\u{001b}[0m";
            }
        }

        out
    }

    pub fn guesses(&self) -> String {
        let mut out = String::new();

        for i in 0..self.correct_guesses.len() {
            let c = &self.correct_guesses[i..i + 1];
            out += "\u{001b}[32m";
            out += c;
            out += "\u{001b}[0m ";
        }

        for i in 0..self.wrong_guesses.len() {
            let c = &self.wrong_guesses[i..i + 1];
            out += "\u{001b}[31m";
            out += c;
            out += "\u{001b}[0m ";
        }

        out.trim_end().to_owned()
    }

    pub fn can_guess(&self) -> bool {
        (self.wrong_guesses.len() as u32) < self.max_guesses
    }

    pub fn guesses_left(&self) -> u32 {
        self.max_guesses - (self.wrong_guesses.len() as u32)
    }

    pub fn has_revealed_word(&self) -> bool {
        for i in 0..self.word.len() {
            let c = &self.word[i..i + 1];
            if !self.correct_guesses.contains(c) {
                return false;
            }
        }

        return false;
    }

    pub fn get_word(&self) -> &String {
        &self.word
    }

    pub fn get_found_count(&self) -> u32 {
        let mut count = 0;
        for i in 0..self.word.len() {
            let c = &self.word[i..i + 1];
            if self.correct_guesses.contains(c) {
                count += 1;
            }
        }

        count
    }


}
